File: imgboxdown.rb
#!/usr/bin/env ruby
#
# Imgbox Downloader
# http://cixtor.com/
# https://github.com/cixtor/mamutools
# http://imgbox.com/
#
# * What is imgbox?
#   Imgbox is a free image hosting service.
# * What file types are supported?
#   Images of the file types JPG, GIF and PNG are supported.
# * What is the maximum file size?
#   Each file must be 10MB or smaller.
# * How long are images stored?
#   Images are stored for life time (see our terms of service).
# * Should I keep a backup of my images on my local machine?
#   Yes, you always should backup your files.
# * I registered for a free account. How many images can I store?
#   There are no limits (fair use).
# * Can I inline link (hotlink) the original images I uploaded?
#   Yes, but please use the provided "full size" share codes only. Other URLs may change in the future.
# * Are there bandwidth limitations for inlin linking (hotlinking)?
#   There is no hard limit. Abusive referrers will be blocked.
# * Why are my thumbnail previews square?
#   Thumbnails on your "My Images" and "My Galleries" pages are displayed as squares to preserve the layout.
#
require 'optparse'
class Imgboxdown
    attr_reader :config, :optparse
    def initialize
        @config = {
            :filename => __FILE__,
            :user_agent => 'Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)',
            :photo_id => nil,
            :gallery_id => nil
        }
        @optparse = OptionParser.new do |option|
            option.on('-i arg', '--id=arg', 'Specify the photo identifier.'){ |arg| config[:photo_id] = arg }
            option.on('-g arg', '--gallery=arg', 'Specify the gallery identifier.'){ |arg| config[:gallery_id] = arg }
        end
        usage
        execute
    end
    def success(message)
        puts "\e[0;92mOK.\e[0m #{message}"
    end
    def highlight(message)
        puts "\e[0;93m[>]\e[0m #{message}"
    end
    def fail(message)
        puts "\e[0;91m[x]\e[0m #{message}"
        exit(1)
    end
    def usage
        puts 'Imgbox Downloader'
        puts '    http://cixtor.com/'
        puts '    https://github.com/cixtor/mamutools'
        puts '    http://imgbox.com/'
        puts
    end
    def get_photo(photo_id, config)
        image_content = %x{curl --silent --user-agent '#{config[:user_agent]}' 'http://imgbox.com/#{photo_id}'}
        image_lines = image_content.split("\n")
        image_lines.each do |image_line|
            if image_match = image_line.chomp.match(/<img alt=".*" class="box" id="img" onclick=".*" src="(.*)" title="(.*)" \/>/) then
                remote_image = image_match[1].gsub('&amp;', '&')
                highlight "Downloading '#{remote_image}' as '#{image_match[2]}'"
                %x{wget --quiet --user-agent='#{config[:user_agent]}' '#{remote_image}' -O '#{image_match[2]}'}
            end
        end
    end
    def execute
        begin
            optparse.parse!
            if config[:photo_id].nil? and config[:gallery_id].nil? then
                raise OptionParser::MissingArgument
            end
            #
            if !config[:photo_id].nil? then
                get_photo(config[:photo_id], config)
            end
            #
            if !config[:gallery_id].nil? then
                gallery_folder = "imgbox-#{config[:gallery_id]}"
                if File.exists?(gallery_folder) then
                    success "Gallery folder already exists: \e[0;93m#{gallery_folder}\e[0m"
                else
                    if Dir.mkdir(gallery_folder) then
                        Dir.chdir(gallery_folder)
                        success "Gallery folder created: \e[0;93m#{gallery_folder}\e[0m"
                    else
                        fail "Could not create gallery folder: \e[0;93m#{gallery_folder}\e[0m"
                    end
                end
                #
                album_content = %x{curl --silent 'http://imgbox.com/g/#{config[:gallery_id]}'}
                album_lines = album_content.split("\n")
                album_lines.each do |line|
                    if match = line.chomp.match(/<a href="\/(.*)" class="gallery_img"><img alt="(.*)" src="(.*)" \/><\/a>/) then
                        get_photo(match[1], config)
                    end
                end
            end
            success 'Finished'
        rescue OptionParser::MissingArgument, OptionParser::InvalidOption
            p optparse
            fail "Process failed, #{$!.to_s}"
        end
    end
end
imgboxdown = Imgboxdown.new
#

File: imgboxdown.go
/**
 * Imgbox Downloader
 * http://cixtor.com/
 * https://github.com/cixtor/mamutools
 * http://imgbox.com/
 *
 * ImgBox is a free image hosting service which supports JPG, GIF and PNG files
 * with a maximum of 10 MB of size, and they can be stored for a life time, they
 * have no limits to store your images once you register for a free account. You
 * can also inline link (hotlink) the images uploaded using the provided full-
 * size share codes only (considering that other URLs may change in the future).
 * There is no hard limit for inline linking (hotlinking) but there are
 * bandwidth limitations in case of abusive referrers where the requests will be
 * blocked.
 */

package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"regexp"
	"strings"
)

var photo_id = flag.String("photo", "", "Specify the photo identifier to download")
var gallery_id = flag.String("gallery", "", "Specify the gallery identifier to download all its images")

const user_agent = "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"

func getRemoteContent(url string) []byte {
	client := &http.Client{}
	req, err1 := http.NewRequest("GET", url, nil)

	if err1 == nil {
		req.Header.Set("User-Agent", user_agent)
		resp, err2 := client.Do(req)

		if err2 == nil {
			defer resp.Body.Close()
			body, _ := ioutil.ReadAll(resp.Body)
			return body
		}
	}

	return nil
}

func getRemoteContentSplitted(url string) []string {
	var body []byte = getRemoteContent(url)
	var str_lines []string = strings.Split(string(body), "\n")

	return str_lines
}

func getPhoto(photo_id string) {
	var r = regexp.MustCompile(`<img alt=".*" class=".*" id="img" onclick=".*" src="(.*)" title="(.*)" \/>`)
	var photo_url string = fmt.Sprintf("http://imgbox.com/%s", photo_id)
	var image_lines []string = getRemoteContentSplitted(photo_url)

	for _, line := range image_lines {
		var results []string = r.FindStringSubmatch(line)

		if len(results) > 0 {
			fmt.Printf("--> Downloading '%s' as '%s'\n", results[1], results[2])
			var filedata []byte = getRemoteContent(results[1])
			ioutil.WriteFile(results[2], filedata, 0644)
		}
	}
}

func getGallery(gallery_id string) {
	var gallery_folder string = fmt.Sprintf("imgbox-%s", gallery_id)
	_, err1 := os.Stat(gallery_folder)

	if err1 == nil {
		fmt.Printf("OK. Gallery folder already exists: %s\n", gallery_folder)
	} else {
		err2 := os.Mkdir(gallery_folder, 0755)

		if err2 == nil {
			fmt.Printf("OK. Gallery folder created: %s\n", gallery_folder)
		} else {
			fmt.Printf("[x] Could not create gallery folder: %s\n", gallery_folder)
			os.Exit(1)
		}
	}

	os.Chdir(gallery_folder)

	var r = regexp.MustCompile(`<a href="\/(.*)"><img alt=".*" src="[a-z]+:\/\/s\.imgbox\.com\/.*" \/><\/a>`)
	var album_url string = fmt.Sprintf("http://imgbox.com/g/%s", gallery_id)
	var album_lines []string = getRemoteContentSplitted(album_url)

	for _, line := range album_lines {
		var results []string = r.FindStringSubmatch(line)

		if len(results) > 0 {
			getPhoto(results[1])
		}
	}
}

func download(photo_id string, gallery_id string) {
	if photo_id != "" {
		fmt.Printf("OK. Download photo: %s\n", photo_id)
		getPhoto(photo_id)
	} else if gallery_id != "" {
		fmt.Printf("OK. Download gallery: %s\n", gallery_id)
		getGallery(gallery_id)
	} else {
		fmt.Printf("Missing argument: -photo or -gallery\n")
		flag.Usage()
		os.Exit(1)
	}
}

func main() {
	flag.Usage = func() {
		fmt.Printf("Imgbox Downloader\n")
		fmt.Printf("  http://cixtor.com/\n")
		fmt.Printf("  https://github.com/cixtor/mamutools\n")
		fmt.Printf("  http://imgbox.com/\n")
		fmt.Printf("Usage:\n")
		flag.PrintDefaults()
	}

	flag.Parse()

	download(*photo_id, *gallery_id)
	fmt.Printf("OK. Finished\n")
}
